src/components/DuplicateCourseModal.js

```diff
@@ -0,0 +1,162 @@
+      locationId: courseListing.locationId,
+      registrarId: courseListing.registrarId,
+      servicepointId: courseListing.servicepointId,
+      termId: term,
+    } }).json();
+
+    for (let i = 0; i < courseListing.instructorObjects.length; i++) {
+      await ky.post(`coursereserves/courselistings/${newCourseListing.id}/instructors`, { json: {
+        courseListingId: newCourseListing.id,
+        name: courseListing.instructorObjects[i].name,
+        userId: courseListing.instructorObjects[i].userId,
+      } }).json();
+    }
+
+    for (let i = 0; i < reserves.length; i++) {
+      await ky.post(`coursereserves/courselistings/${newCourseListing.id}/reserves`, { json: {
+        courseListingId: newCourseListing.id,
+        itemId: reserves[i].itemId,
+      } }).json();
+    }
+
+    const newCourse = await ky.post(`coursereserves/courselistings/${newCourseListing.id}/courses`, { json: {
+      courseListingId: newCourseListing.id,
+      courseNumber: course.courseNumber,
+      departmentId: course.departmentId,
+      description: course.description,
+      name: `${course.name} - ${intl.formatMessage({ id: 'ui-courses.button.duplicate' })}`,
+      sectionName: course.sectionName,
+    } }).json();
+
+    if (duplicateCrosslistedCourses) {
+      for (let i = 0; i < crossListed.length; i++) {
+        await ky.post(`coursereserves/courselistings/${newCourseListing.id}/courses`, { json: {
+          courseListingId: newCourseListing.id,
+          courseNumber: crossListed[i].courseNumber,
+          departmentId: crossListed[i].departmentId,
+          description: crossListed[i].description,
+          name: `${crossListed[i].name} - ${intl.formatMessage({ id: 'ui-courses.button.duplicate' })}`,
+          sectionName: crossListed[i].sectionName,
+        } }).json();
+      }
+    }
+
+    history.push(`/cr/courses/${newCourse.id}`);
+
+    onClose();
+  };
+
+  return (
+    <Modal
+      dismissible
+      id="duplicate-course-modal"
+      label={<FormattedMessage id="ui-courses.duplicateCourse" />}
+      onClose={onClose}
+      open
+    >
+      <Select
+        autoFocus
+        dataOptions={terms}
+        fullWidth
+        data-test-select-term-for-duplicate-course
+        label={<FormattedMessage id="ui-courses.field.term" />}
+        onChange={e => setTerm(e.target.value)}
+        required
+        value={term}
+      />
+      <Checkbox
+        checked={duplicateCrosslistedCourses}
+        data-test-checkbox-duplicate-crosslisted-courses
+        label={<FormattedMessage id="ui-courses.duplicateCourse.duplicateCrosslistedCourses" />}
+        onChange={e => setDuplicateCrosslistedCourses(e.target.checked)}
+      />
+      <Button
+        disabled={term === undefined}
+        fullWidth
+        data-test-duplicate-modal-clickable-duplicate
+        onClick={duplicateCourse}
+      >
+        <FormattedMessage id="ui-courses.button.createDuplicateCourses" />
+      </Button>
+    </Modal>
```

Comment:
Holy crap, this code is clean. It's a a joy to read. A really nice example of application code that works with the grain of the underlying libraries rather than fighting against them (as I admit I often feel like I've ended up doing). Beautiful.